### PR TITLE
address issue #1

### PR DESCRIPTION
**Pull request by rflorenc**
@bakito 
@radudd 

https://gist.github.com/rflorenc/c150c2a368bc39412e89a1e0c77d12a9

Will test it on a cluster without velero.io this week.

### CODE REVIEW

File: controllers/drwatcher_controller.go [hunk 1]
```diff
@@ -90,6 +95,18 @@ func (r *DRWatcherReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	return ctrl.Result{}, nil
 }
 
+func (r *DRWatcherReconciler) preRequisites() error {
+	for _, unstructuredCrd := range veleroInstall.AllCRDs().Items {
+		crd := &apiextv1beta1.CustomResourceDefinition{}
+		err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstructuredCrd.Object, crd)
```
Comment (snorwin) [inline controllers/drwatcher_controller.go]: How this section checks that the velero CRDs are in fact installed on the cluster?

File: controllers/drwatcher_controller.go [hunk 2]
```diff
@@ -36,6 +38,11 @@ func (r *DRWatcherReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	var err error
 	var drwatcherCR drv1.DRWatcher
 
+	err = r.preRequisites()
+	if err != nil {
+		return ctrl.Result{}, err
```
Comment (snorwin) [inline controllers/drwatcher_controller.go]: Should it loop the requests until the CRDs are installed? Does it make sense to check the perquisites in every reconcile?

File: controllers/drwatcher_controller.go (hunk 1 above)
Comment (rflorenc) [reply controllers/drwatcher_controller.go]: This loop checks that those crds are registered in the velero gvk. It is just unmarshalling the objects before any conversion or updates are done.

If we start the controller without the prerequisite velero.io group, for example in a fresh cluster where velero was never installed, we'll get errors similar to. 
2021-03-19T10:39:04.916Z ERROR controller-runtime.source if kind is a CRD, it should be installed before calling Start {"kind": "Backup.velero.io", "error": "no matches for kind \"Backup\" in version \"velero.io/v1\""}

File: controllers/drwatcher_controller.go (hunk 2 above)
Comment (rflorenc) [reply controllers/drwatcher_controller.go]: I also thought about making this a one time check, for example at main.go, before initializing the manager etc, but not sure here to be honest. If the main infra velero and all its crds/crs are deleted after that one time check the controller will just error out in a loop with errors similar to previous comment.

Can also make it an interval check, like every x minutes and handle errors if not.

**Comment by rflorenc** [comment]
https://gist.github.com/rflorenc/999801b3675b713946ad5b615a19e0df

File: main.go [hunk 3]
```diff
@@ -38,6 +45,26 @@ var (
...
+		crd := &apiextv1beta1.CustomResourceDefinition{}
+		err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstructuredCrd.Object, crd)
+		if err != nil {
+			logger.Error(err, "Required velero.io group CRDs could not be converted.")
+			return err
+		}
+		err = client.Get(ctx, types.NamespacedName{Name: crd.ObjectMeta.Name}, crd)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				logger.Info(fmt.Sprintf("client.Get(): Required velero.io CRD %v not found", crd.ObjectMeta.Name))
```
Comment (snorwin) [inline main.go]: maybe you should return here an error too

File: main.go [hunk 4]
```diff
@@ -38,6 +45,26 @@ var (
 	metricsPort string = ":8080"
 )
 
+func checkPreRequisites(ctx context.Context, client client.Client, logger logr.Logger) error {
+	for _, unstructuredCrd := range veleroInstall.AllCRDs().Items {
+		crd := &apiextv1beta1.CustomResourceDefinition{}
+		err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstructuredCrd.Object, crd)
```
Comment (snorwin) [inline main.go:51]: you could also convert the entire object to  a `CustomResourceDefinitionList`

File: main.go (hunk 3 above)
Comment (snorwin) [inline main.go]: do you really want to log this as Info?

File: controllers/drwatcher_controller.go [hunk 5]
```diff
@@ -95,6 +93,6 @@ func (r *DRWatcherReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&drv1.DRWatcher{}).
 		Owns(&corev1.Pod{}).
-		Owns(&velerov1.Backup{}).
+		// Owns(&velerov1.Backup{}).
```
Comment (snorwin) [inline controllers/drwatcher_controller.go]: debug code

Comment (rflorenc) [reply controllers/drwatcher_controller.go]: This needs to be removed for two reasons:

- we can't set ownership on a non existent Kind (velero isn't present/velero crds were never installed)
- the main infra velero should own this, not the drwatcher.

I didn't see this error while testing on clusters with full velero setup, but on a fresh cluster these issues surface.

**Pull request merged**